shopinvader_auth_jwt/controllers/main.py

```diff
@@ -0,0 +1,91 @@
+    _collection_name = "shopinvader.backend"
+    _default_auth = "jwt_shopinvader"
+
+    @route(
+        ["/shopinvader/<service>/<int:_id>/download"],
+        methods=["GET", "OPTIONS"],
+        auth=_default_auth,
+    )
+    def service_download(self, service, _id=None, **params):
+        params["id"] = _id
+        return self._process_method(service, "download", _id, params=params)
+
+    @classmethod
+    def _get_shopinvader_partner_from_jwt(cls):
+        partner_model = request.env["shopinvader.partner"]
+        partner_email = request.jwt_payload.get("email")
+        backend = cls._get_shopinvader_backend_from_request()
+        if partner_email:
+            shopinvader_partner = cls._find_partner(backend, partner_email)
+            if len(shopinvader_partner) == 1:
+                cls._validate_partner(backend, shopinvader_partner)
+                return shopinvader_partner
+            else:
+                _logger.warning("Wrong HTTP_PARTNER_EMAIL, header ignored")
+                if len(shopinvader_partner) > 1:
+                    _logger.warning(
+                        "More than one shopinvader.partner found for:"
+                        " backend_id={} email={}".format(backend.id, partner_email)
+                    )
+                # Could be because the email is not related to a partner or
+                # because the partner is inactive
+                raise MissingError(_("The given partner is not found!"))
+        return partner_model.browse([])
+
+    @classmethod
+    def _find_partner(cls, backend, partner_email):
+        partner_domain = [
+            ("partner_email", "=", partner_email),
+            ("backend_id", "=", backend.id),
+        ]
+        return request.env["shopinvader.partner"].search(partner_domain, limit=2)
+
+    @classmethod
+    def _validate_partner(cls, backend, partner):
+        with backend.work_on("res.partner") as work:
+            validator = work.component(usage="partner.validator")
+            validator.validate_partner(partner)
+
+    @classmethod
+    def _get_shopinvader_backend_from_request(cls):
+        backend_model = request.env["shopinvader.backend"]
+        return backend_model._get_from_jwt_aud(request.jwt_payload.get("aud"))
+
+    def _get_component_context(self):
+        """Retrieve service component work context.
+
+        Main keys to look for:
+
+        * partner_user: the partner of the current user (matching via email)
+        * partner: every service will use it as the current partner (eg: customer info)
+        * cart_id: sale order ID used to keep in sync client and backend
+        * shopinvader_backend: current shopinvader backend (matching API key)
+        """
+        res = super(InvaderController, self)._get_component_context()
+        res["shopinvader_backend"] = self._get_shopinvader_backend_from_request()
+        # TODO: all services should rely on shopinvader partner
+        # rather than the real partner. SBI: why?
```

Comment:
because if you don't have a shopinvader.partner you are not shopinvader user.

Reply:
@simahawk sure but in the business logic (is that what you mean with "all services" in that comment) we always need the `res.partner`. The `shopinvader.partner` is a technical mean to locate the `res.partner`, and most services should not care about it once the res.partner has been located for them.

Reply:
My point is that when you have the shopinvader.partner you also have the real partner. Not the contrary. 
Also, some services might require specific values which depend on the backend (thus on the binding).

Probably we should simply have something like:
```py
@property
def partner(self):
    return self.work.invader_partner.record_id
```
and rely on the fact that the work context must always have the invader partner.

When I wrote that comment we didn't have the unified way to load the ctx, hence it was easy to miss one of the ctx vars or have them misaligned. Today we could probably get rid of it and maybe implement the change above.

Reply:
Note that I some situations, we may not *need* a `shopinvader.partner`. For instance if we have an external authentication provider (e.g. keycloak), we can use the same customer identity for several shopinvader frontends (why do we call that backends? :/). We have such requirements for a customer that has several web sites, but want a single customer login. And in that case `shopinvader.partner` is a useless overhead.